onnx/shape_inference/implementation.cc

```diff
@@ -115,6 +115,39 @@ void checkShapesAndTypes(const TypeProto& inferredType, const TypeProto& existin
     fail_type_inference("type case unsupported. existing=", existingTypeCase, " inferred=", inferredTypeCase);
   }
 }
+// TypeProto_Tensor or TypeProto_SparseTensor
+template <typename T>
+void generateSymbolicShape(T inferredType, T* symbolicInferredType, SymbolicShape symbolicShape) {
+  if (!inferredType.has_shape()) {
+    *symbolicInferredType = inferredType;
+    return;
+  }
+  for (int i = 0; i < inferredType.shape().dim_size(); ++i) {
+    const auto& inferredDim = inferredType.shape().dim(i);
+    auto* typeAfterSymbolicShapeDim = symbolicInferredType->mutable_shape()->mutable_dim(i);
+    *typeAfterSymbolicShapeDim = inferredDim;
+    // If shape cannot be inferred; set it as a symbolic shape
+    if (inferredType.shape().dim(i).dim_value() == 0 && !inferredType.shape().dim(i).has_dim_param()) {
+      typeAfterSymbolicShapeDim->set_dim_param(symbolicShape.createNew());
+    }
+  }
+}
+
+const TypeProto* enableSymbolicShape(const TypeProto* inferredType, SymbolicShape symbolicShape) {
+  const auto inferred_val_case = inferredType->value_case();
+  TypeProto* typeProto = const_cast<TypeProto*>(inferredType);
```

Comment:
This is odd. I would make the parameter non-const here (and make this the caller's issue).

Reply:
Makes sense. Updated. Thanks